src/main/java/org/springframework/data/elasticsearch/core/IndexCoordinates.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.elasticsearch.core;
+
+import java.util.Arrays;
+
+import org.springframework.util.Assert;
+
+/**
+ * Value object encapsulating index name(s) and index type(s).
+ *
+ * @author Mark Paluch
+ * @author Christoph Strobl
+ * @author Peter-Josef Meisch
+ * @since 4.0
+ */
+public class IndexCoordinates {
+
+	private final String[] indexNames;
+	private String[] typeNames = new String[0];
+
+	public static IndexCoordinates of(String... indexNames) {
+		Assert.notNull(indexNames,"indexNames must not be null");
+		return new IndexCoordinates(indexNames);
+	}
+
+	private IndexCoordinates(String[] indexNames) {
+		Assert.notEmpty(indexNames, "indexNames may not be null or empty");
+		this.indexNames = indexNames;
+	}
+
+	public IndexCoordinates withTypes(String... typeNames) {
+		Assert.notNull(typeNames, "typeNames must not be null");
+		this.typeNames = typeNames;
+		return this;
+	}
+
+	public String getIndexName() {
+		return indexNames != null && indexNames.length > 0 ? indexNames[0] : null;
+	}
+
+	public String[] getIndexNames() {
```

Comment:
Not sure, we might want to return a copy of the array to prevent mutation (e.g `getIndexNames()[0] = "foo"`)